### PR TITLE
Add SDW support for IPC4

### PR DESCRIPTION
**Pull request by bardliao**
We need to do some cleanup especially ipc4-alh.h.

### CODE REVIEW

File: sound/soc/sof/ipc4-alh.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+struct sof_gtw_attributes {
+	// Gateway data requested in low power memory.
+	uint32_t lp_buffer_alloc : 1;
+
+	//Gateway data requested in register file memory.
+	uint32_t alloc_from_reg_file : 1;
+
+	// Reserved field
+	uint32_t _rsvd : 30;
+} __packed;
```
Comment (plbossart) [inline sound/soc/sof/ipc4-alh.h]: this is a common definition if I am not mistaken, it shouldn't be part of ALH support only.

File: sound/soc/sof/ipc4-alh.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+struct AlhMultiGtwCfg {
```
Comment (plbossart) [inline sound/soc/sof/ipc4-alh.h]: Camel Case?

File: sound/soc/sof/ipc4-alh.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+		uint32_t channel_mask;
+	} mapping[1]; //!< Mapping items
+} __packed;
+
+struct sof_alh_configuration_blob {
+	struct sof_gtw_attributes gw_attr;
+	struct AlhMultiGtwCfg alh_cfg;
+};
+
+
```
Comment (plbossart) [inline sound/soc/sof/ipc4-alh.h]: spurious line

File: sound/soc/sof/ipc4-topology.c [hunk 4]
```diff
@@ -1035,6 +1036,19 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
...
 			copier_data->gtw_cfg.config_length >>= 2;
 			break;
+		case SOF_DAI_INTEL_ALH:
+		{
+			struct sof_alh_configuration_blob *blob;
+
+			blob = devm_kzalloc(sdev->dev, sizeof(*blob), GFP_KERNEL);
+			blob->alh_cfg.count = 1;
+			blob->alh_cfg.mapping[0].alh_id = copier_data->gtw_cfg.node_id;
+			blob->alh_cfg.mapping[0].channel_mask = 0x3;
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c]: shouldn't this come from topology? It's not clear to me why this is stereo only

File: sound/soc/sof/ipc4-alh.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+#define SNDW_MULTI_GTW_BASE  0x50
```
Comment (plbossart) [inline sound/soc/sof/ipc4-alh.h]: a comment on what this is wouldn't hurt. Not sure if this is a sizeof() or a register offset.

File: drivers/soundwire/intel.c [hunk 6]
```diff
@@ -1005,9 +1005,13 @@ static int intel_trigger(struct snd_pcm_substream *substream, int cmd, struct sn
 {
 	struct sdw_cdns *cdns = snd_soc_dai_get_drvdata(dai);
 	struct sdw_intel *sdw = cdns_to_intel(cdns);
+	struct sdw_intel_link_res *res = sdw->link_res;
 	struct sdw_cdns_dma_data *dma;
 	int ret = 0;
 
+	if (res->ops && res->ops->trigger)
+		res->ops->trigger(substream, cmd, dai);
+
```
Comment (plbossart) [inline drivers/soundwire/intel.c:1014]: it's not super-elegant, we already have an intel_free_stream() below to take care of the TRIGGER_SUSPEND case.

File: sound/soc/sof/intel/hda-dai.c [hunk 7]
```diff
@@ -626,6 +626,7 @@ static int ipc4_be_dai_trigger(struct snd_pcm_substream *substream,
 		struct snd_sof_widget *pipe_widget = swidget->pipe_widget;
 		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
 
+		sdev = snd_soc_component_get_drvdata(swidget->scomp);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai.c:628]: You probably want a comment to explain the change - likely that the ALH/SoundWire dais are registered by different components for each link.

File: sound/soc/sof/intel/hda.c [hunk 8]
```diff
@@ -147,9 +147,21 @@ static int sdw_free_stream(struct device *dev,
...
 
+static int sdw_trigger(struct snd_pcm_substream *substream, int cmd, struct snd_soc_dai *dai)
+{
+	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(dai, substream->stream);
+	struct snd_sof_widget *swidget = w->dobj.private;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(swidget->scomp);
+
+	if (sdev->pdata->ipc_type != SOF_INTEL_IPC4)
+		return 0;
+	return ipc4_be_dai_trigger(substream, cmd, dai);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: it's not super elegant to have tests for IPC4. we could make the sdw_callback non-const or something where we only do the test on initialization.

Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: Or we do more surgery and move all the DAI stuff back in the sound/soc/sof side.

File: drivers/soundwire/intel.c (hunk 6 above)
Comment (bardliao) [reply drivers/soundwire/intel.c:1014]: > it's not super-elegant, we already have an intel_free_stream() below to take care of the TRIGGER_SUSPEND case.

But intel_free_stream() is only for suspend. What we do in the trigger ops is for trigger suspend, stop, and pause. It seems that intel_free_stream() doesn't cover the pause case.

File: sound/soc/sof/intel/hda-dai.c (hunk 7 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda-dai.c:628]: > You probably want a comment to explain the change - likely that the ALH/SoundWire dais are registered by different components for each link.

Commit message updated.

File: sound/soc/sof/ipc4-topology.c [hunk 9]
```diff
@@ -1035,6 +1036,18 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 			/* at this point amount of dwords */
 			copier_data->gtw_cfg.config_length >>= 2;
 			break;
+		case SOF_DAI_INTEL_ALH:
+		{
+			struct sof_ipc4_alh_configuration_blob *blob;
+
+			blob = devm_kzalloc(sdev->dev, sizeof(*blob), GFP_KERNEL);
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: can we use kzalloc() and free in unprepare()?

File: sound/soc/sof/ipc4-topology.c [hunk 10]
```diff
@@ -1035,6 +1036,18 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
...
+			blob->alh_cfg.count = 1;
+			blob->alh_cfg.mapping[0].alh_id = copier_data->gtw_cfg.node_id;
+			blob->gw_attr.lp_buffer_alloc = 0;
+			copier_data->gtw_cfg.config_length = sizeof(*blob) >> 2;
+			ipc4_copier->copier_config = (uint32_t *)blob;
+			break;
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: can we unify all the prepare code for ALH after the call to init_audio_format()?

File: sound/soc/sof/ipc4-topology.c [hunk 11]
```diff
@@ -1054,6 +1067,32 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
...
+			blob = (struct sof_ipc4_alh_configuration_blob *)&ipc4_copier->copier_config;
+			ch_map = copier_data->base_config.audio_fmt.ch_map;
+			for (i = 0; ch_map > 0; i++) {
+				if ((ch_map & 0xf) != 0xf)
+					 blob->alh_cfg.mapping[0].channel_mask |= BIT(i);
+				ch_map >>= 4;
+			}
+			break;
+		}
+		break;
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: default: break?

File: sound/soc/sof/ipc4-topology.c (hunk 9 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c]: > can we use kzalloc() and free in unprepare()?

In fact, blob is useless after 
```
memcpy(*ipc_config_data + sizeof(*copier_data),
                       *data, copier_data->gtw_cfg.config_length * 4);
```
So I will free after it.

File: sound/soc/sof/ipc4-topology.c (hunk 10 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c]: > can we unify all the prepare code for ALH after the call to init_audio_format()?

Yes, done.

File: sound/soc/sof/ipc4-topology.c (hunk 11 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c]: > default: break?

I am not sure if a default case is needed or not. I will remove the redundant break.

File: sound/soc/sof/ipc4-alh.h [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ *
+ * Author: Bard Liao <yung-chuan.liao@linux.intel.com>
+ */
+
+#ifndef __SOF_INTEL_IPC4_ALH_H
+#define __SOF_INTEL_IPC4_ALH_H
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-alh.h]: should this be:
__SOUND_SOC_SOF_IPC4_ALH_H
?

File: sound/soc/sof/ipc4-alh.h [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "ipc4-topology.h"
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-alh.h]: why not put this into ipc4-topology.h directly?

File: drivers/soundwire/intel.c [hunk 14]
```diff
@@ -1005,9 +1005,13 @@ static int intel_trigger(struct snd_pcm_substream *substream, int cmd, struct sn
 {
 	struct sdw_cdns *cdns = snd_soc_dai_get_drvdata(dai);
 	struct sdw_intel *sdw = cdns_to_intel(cdns);
+	struct sdw_intel_link_res *res = sdw->link_res;
 	struct sdw_cdns_dma_data *dma;
 	int ret = 0;
 
+	if (res->ops && res->ops->trigger)
+		res->ops->trigger(substream, cmd, dai);
```
Comment (ujfalusi) [inline drivers/soundwire/intel.c:1013]: Should this be done after the `snd_soc_dai_get_dma_data()` ?

File: drivers/soundwire/intel.c (hunk 6 above)
Comment (ujfalusi) [reply drivers/soundwire/intel.c:1014]: free_stream ops  is pretty specific to free a stream and it needs some specific parameters, I'm not sure if that could be handled via this new callback also.

File: include/linux/soundwire/sdw_intel.h [hunk 15]
```diff
@@ -119,6 +119,8 @@ struct sdw_intel_ops {
 			     struct sdw_intel_stream_params_data *params_data);
 	int (*free_stream)(struct device *dev,
 			   struct sdw_intel_stream_free_data *free_data);
+	int (*trigger)(struct snd_pcm_substream *substream,
+		       int cmd, struct snd_soc_dai *dai);
```
Comment (ujfalusi) [inline include/linux/soundwire/sdw_intel.h:123]: other callbacks have device + some sort of struct as parameter, should this be aligned as well?

File: sound/soc/sof/intel/hda-dai.c (hunk 7 above)
Comment (ujfalusi) [reply sound/soc/sof/intel/hda-dai.c:628]: Hrm, but we can have only _one_ sdev, no?

File: sound/soc/sof/intel/hda.c [hunk 16]
```diff
@@ -185,6 +185,10 @@ static int hda_sdw_probe(struct snd_sof_dev *sdev)
 
 	memset(&res, 0, sizeof(res));
 
+	/* sdw trigger callback is only needed for IPC4 */
+	if (sdev->pdata->ipc_type == SOF_INTEL_IPC4)
+		sdw_callback.trigger = ipc4_be_dai_trigger;
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda.c:190]: I see, we are re-using the existing `ipc4_be_dai_trigger()`, that's why the different sets of parameters for the trigger callback.

File: sound/soc/sof/intel/hda-dai.c [hunk 17]
```diff
@@ -610,8 +610,8 @@ static const struct snd_soc_dai_ops ipc3_ssp_dai_ops = {
 	.shutdown = ssp_dai_shutdown,
 };
 
-static int ipc4_be_dai_trigger(struct snd_pcm_substream *substream,
-			       int cmd, struct snd_soc_dai *dai)
+int ipc4_be_dai_trigger(struct snd_pcm_substream *substream,
+			int cmd, struct snd_soc_dai *dai)
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai.c]: 'int cmd' one line up?

File: sound/soc/sof/intel/hda.h [hunk 18]
```diff
@@ -774,5 +774,8 @@ irqreturn_t cnl_ipc4_irq_thread(int irq, void *context);
 int cnl_ipc4_send_msg(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg);
 irqreturn_t hda_dsp_ipc4_irq_thread(int irq, void *context);
 int hda_dsp_ipc4_send_msg(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg);
+int ipc4_be_dai_trigger(struct snd_pcm_substream *substream,
+			int cmd, struct snd_soc_dai *dai);
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda.h]: the 'int cmd,' can go one line up.

**Review by ujfalusi — CHANGES_REQUESTED** [review]

File: sound/soc/sof/ipc4-topology.c [hunk 19]
```diff
@@ -1054,6 +1055,41 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
...
+					 blob->alh_cfg.mapping[0].channel_mask |= BIT(i);
+				ch_map >>= 4;
+			}
+
+			ipc4_copier->copier_config = (uint32_t *)blob;
+			break;
+		}
+		}
+	}
+	}
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:1090]: indentation issues here?

File: sound/soc/sof/ipc4-topology.c [hunk 20]
```diff
@@ -1099,6 +1135,10 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		memcpy(*ipc_config_data + sizeof(*copier_data),
 		       *data, copier_data->gtw_cfg.config_length * 4);
 
+	/* copier_config is useless after memcpy */
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c]: is no longer needed

File: drivers/soundwire/intel.c (hunk 14 above)
Comment (bardliao) [reply drivers/soundwire/intel.c:1013]: > Should this be done after the `snd_soc_dai_get_dma_data()` ?

Given that 1. the trigger ops doesn't need the dma data, and 2. the trigger ops is used to send IPC to firmware. I think we should call the trigger ops no matter we get dma data successfully or not.

File: sound/soc/sof/intel/hda-dai.c (hunk 7 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda-dai.c:628]: > Hrm, but we can have only _one_ sdev, no?

Yes, and we can get the only one sdev from swidget->scomp

File: sound/soc/sof/ipc4-topology.c (hunk 19 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1090]: just an extra curly braces outside the inner switch/case

**Comment by ranj063** [comment]
@bardliao I'm going to merge this for now to enable testing in CI. Can you please work on the outstanding comments as fixups.

**Pull request merged**